File: AdventOfCode2019/06/day06.js
var input = require("./input.js");

const orbitsToDictionary = (orbits) => {
  const out = {};
  orbits.forEach((orbit) => {
    const [left, right] = orbit;
    out[left] = out[left] || [];
    out[left].push(right);
  });

  return out;
};

const buildPaths = (orbitsDict, x, path = [], out = {}) => {
  out[x] = out[x] || [];
  out[x] = path;

  const directOrbits = orbitsDict[x];
  if (directOrbits) {
    directOrbits.forEach((orbit) => {
      buildPaths(orbitsDict, orbit, [...path, x], out);
    });
  }

  return out;
};

// Part 1
const dict = orbitsToDictionary(input.orbits);
const paths = buildPaths(dict, "COM");
const part1 = Object.keys(paths).reduce((total, curr) => total + paths[curr].length, 0);
console.log(part1);

// Part 2
const youPath = paths["YOU"];
const santaPath = paths["SAN"];
let commonNodesCount = 0;
while (youPath[commonNodesCount] == santaPath[commonNodesCount]) {
  commonNodesCount++;
}

const youToSantaLength = youPath.length - commonNodesCount + santaPath.length - commonNodesCount;
console.log(youToSantaLength);

File: AdventOfCode2019/06/Day06.hs
module Main where

import qualified Data.Map as Map
import Input (orbits)

toMap = Map.fromListWith (++) . map (\(x, y) -> (x, [y]))

paths :: [Char] -> [[Char]] -> Map.Map [Char] [[Char]] -> Map.Map [Char] [[Char]] -> Map.Map [Char] [[Char]]
paths x path orbits out = result
  where
    directOrbits = Map.lookup x orbits
    withCurrent = Map.insert x path out
    result = case directOrbits of
      Nothing -> withCurrent
      Just connections -> foldr (\y acc -> paths y (x : path) orbits acc) withCurrent connections

-- Part 1
orbitsMap = toMap orbits

distances = paths "COM" [] orbitsMap Map.empty

part1 = sum $ map length $ Map.elems distances

-- Part 2
pathTo x m = case Map.lookup x m of
  Just p -> reverse (x : p)
  Nothing -> error "No path"

youPath = pathTo "YOU" distances

santaPath = pathTo "SAN" distances

aToBPath s a@(x : xs) b@(y : ys) = if x == y then aToBPath x xs ys else reverse a ++ [s] ++ b
aToBPath _ _ _ = error "No common"

youToSanta = aToBPath "" youPath santaPath

part2 = length youToSanta - 2 - 1

main = do
  print part1
  print part2